### PR TITLE
Bugfix: Reading Weighted Graph with Floating Point Weight Values 

### PR DESCRIPTION
**Pull request by c-bebop**
This PR addresses the issue that so far GDSB was not capable of reading in weights that are actually floating point values in the graph file.

### CODE REVIEW

File: include/gdsb/graph_input.h [hunk 1]
```diff
@@ -130,7 +139,7 @@ std::tuple<Vertex, uint64_t> read_graph(std::istream& input,
         if constexpr (GraphParameters::is_weighted())
         {
             string_source = string_position;
-            w = read_ulong(string_source, &string_position);
+            w = read_float(string_source, &string_position);
```
Comment (bernlu) [inline include/gdsb/graph_input.h:142]: does this mean that all graphs are now read as if they have floating point weights? Seems like `read_ulong` is no longer used anywhere then. Maybe remove that code?

File: src/graph_input.cpp [hunk 2]
```diff
@@ -26,6 +26,20 @@ unsigned long read_ulong(char const* source, char** end)
     return value;
 }
 
+float read_float(char const* source, char** end)
+{
+    float const value = std::strtof(source, end);
+
+#if !defined(__clang__)
+    if (errno == ERANGE)
```
Comment (bernlu) [inline src/graph_input.cpp:34]: should errno be set to 0 first, to make sure that we do not check for an old error?

File: src/graph_input.cpp [hunk 3]
```diff
@@ -26,6 +26,20 @@ unsigned long read_ulong(char const* source, char** end)
     return value;
 }
 
+float read_float(char const* source, char** end)
+{
+    float const value = std::strtof(source, end);
+
+#if !defined(__clang__)
```
Comment (bernlu) [inline src/graph_input.cpp:33]: what happens when clang is used? I think a comment to explain this code would be useful.

File: include/gdsb/graph_input.h [hunk 4]
```diff
@@ -22,6 +22,15 @@ namespace gdsb
 //! to read next from for subsequent reads of integers within one string (stream).
 unsigned long read_ulong(char const* source, char** end = nullptr);
 
+//! Reads in space separated floating point values. Returns infinity if value at
+//! position source can not be interpreted as a number.
+//! @param source   Points to the beginning of the character (sub string) to be
+//! interpreted.
+//! @param end      Points to nullptr on first call, will (point) out position
```
Comment (bernlu) [inline include/gdsb/graph_input.h:29]: this nullptr description is slightly different to the documentation of `std::strtof` - there i says that the end pointer will be ignored if it is a nullptr.

**Review by bernlu — APPROVED** [review]
left some comments on code I am not sure about.

File: src/graph_input.cpp (hunk 2 above)
Comment (c-bebop) [reply src/graph_input.cpp:34]: I've never seen anyone setting `errno` a-priori :thinking: The [cppreference](https://en.cppreference.com/w/cpp/error/errno) entry does not explain if it is good practice to set it to `0` after (not before) the comparison with `ERANGE` has succeeded.

File: include/gdsb/graph_input.h (hunk 4 above)
Comment (c-bebop) [reply include/gdsb/graph_input.h:29]: I think this is actually in line with what my intention is. If one calls this function for the first time one simply uses `std::nullptr` as `end` which then will be ignored. But it should still return the out position to provide as the next value for `source` if `read_float()` is called consecutively to read multiple floats within one string.

File: src/graph_input.cpp (hunk 2 above)
Comment (c-bebop) [reply src/graph_input.cpp:40]: `errno` is set to `0` now after it was set to `ERANGE`.

File: src/graph_input.cpp (hunk 3 above)
Comment (c-bebop) [reply src/graph_input.cpp:39]: I added a comment concerning the preprocessing pragma checking if the clang compiler is used.

File: include/gdsb/graph_input.h (hunk 4 above)
Comment (c-bebop) [reply include/gdsb/graph_input.h]: Revised the documentation.

**Comment by c-bebop** [comment]
Thanks for the review!

**Pull request merged**